Parse formatted number string to number

praveentiru opened the pull request:
This PR fixes #35. It is a draft PR to implement new feature to convert a formatted string to Maybe Float.

There is still some work in docs. Please review the approach and let me know your thoughts.

praveentiru committed:
New feature: Parse formatted number string to number value based on locale

praveentiru committed:
Change summary: Add new feature to parse formatted string to number
value based on locale
1. Add new function parse to expose new feature
2. Add new function parseString in Parser.elm to implement parse
3. Add new function joinParts to join the number parts into float

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,40 @@ parse locale original =
+    let
+        isNegative : Bool
+        isNegative = 
+            if String.left 1 value == locale.negativePrefix then
```
This is probably not the right approach because:

1. `Locale.negativePrefix` is a string and it is not guaranteed to be 1 character long, one could have, for example `" -"`, for example; the numbers `"-123"` and `" -123"` would both be incorrectly labeled as positive
2. We have  `Locale.negativeSuffix`, so one can have `""` as a prefix and `"-"` as a suffix, for example; the number `"123-` would be incorrectly interpreted as a positive number

praveentiru replied:
We can use the following regex pattern to match negative number by using both prefix and suffix. But, we will run into problem when Prefix or, suffix are brackets (either ( or, [) as they have special regex meaning. We would have to check for them and handle them appropriately.

`^locale.negativePrefix.*locale.negativeSuffix$`
Ex: `^.*-$`

praveentiru replied:
Updated this using `String.startsWith` and `String.endsWith` to match prefix and suffix.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,40 @@ parse locale original =
+                True
+            else
+                False
+        chrs : Regex.Regex
+        chrs = 
+            ["[", locale.negativePrefix, locale.negativeSuffix, locale.thousandSeparator, "]"]
+                |> String.concat
+                |> Regex.fromString
+                |> Maybe.withDefault Regex.never
```
I think this strategy might be overcomplicated and potentially wrong since `negativePrefix`, `negativeSuffix` and `thousandSeparatot` might
1. compromise the regular express syntax and
2.  be misleading if longer than 1 character.

What I would try is:
- Split the number by `decimalSeparator`, i.e. `"-3,141.5" => ("-3,141", "5")`
- Clean up the parts, i.e. `("-3,141", "5") => ("3141", "5")`
- Parse the parts to `Int`, i.e. `("3141", "5") => (3141, 5)`
- Sum the integers digits to the decimal digits corrected by the right factor, i.e. `(3141, 5) => 3141 + 0.5 => 3141.5`

praveentiru replied:
I am using above regex to clean the string of symbols. I can achieve this by flipping the regex to filter everything that is not a digit. Let me give it a try.

cuducos replied:
> I am using above regex to clean the string of symbols. I can achieve this by flipping the regex to filter everything that is not a digit.

This ignores my point on the risk of using strigs that we don't know beforehand to build a regular expression. If a user decides that `"["` is the prefix, the resulting regular expression is invalid, for example.

I wouldn't go down this rabbit hole if we have an alternative that does not require building a regular expression with unknown strings.

praveentiru replied:
I am using the `[^\d]` regex to match all characters that are not digits. I am using this regex to clean the string after splitting them. But, in terms of purity while parsing `Locale.thousandSeparator` has no value anymore. Refer the second test for parse function in the new commit.

Let me know your thoughts.

praveentiru replied:
Change implemented.

cuducos reviewed (changes requested):
Once more, many thanks for your extra hand here :tada: 

I made some comments to try to get to a more flexible implementation — what do you think?

praveentiru committed:
Simplify logic as per feedback

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,61 @@ parse locale original =
+            value
+                |> String.endsWith locale.negativeSuffix
+
+        chrs : Regex.Regex
+        chrs =
+            "[^\\d]"
+                |> Regex.fromString
+                |> Maybe.withDefault Regex.never
+
+        splitValue : String -> List String
+        splitValue nbr =
+            nbr
+                |> String.split locale.decimalSeparator
+                |> List.map (Regex.replace chrs (\_ -> ""))
```
`chrs` and `nbrs` are not readable and in Elm we usually try to use meaningful variable names. Also, I don't think it character or number describes what they are. I think we can wrap this logic in more meaningful abstractions and names (this is just pseudo-code as an example, I haven't run this code):

```suggestion
         notDigits : Regex.Regex
         notDigits =
             "\\D"
                 |> Regex.fromString
                 |> Maybe.withDefault Regex.never

        onlyDigits : String -> String
        onlyDigits =
            Regex.replace notDigits (\_ -> "")
            
        splitDecimals : String -> List String
        splitDecimals value =
           value
                |> String.split locale.decimalSeparator
                |> List.map onlyDigits
```

praveentiru replied:
I am coming from Go coding background. It is notional to provide small names such as r, s, and i. Corrected the code as per advise. Thanks for feedback.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,61 @@ parse locale original =
+    let
+        isNegative : Bool
+        isNegative =
+            negativePrefixMatch && negativeSuffixMatch
+
+        negativePrefixMatch : Bool
+        negativePrefixMatch =
+            value
+                |> String.startsWith locale.negativePrefix
+
+        negativeSuffixMatch : Bool
+        negativeSuffixMatch =
+            value
+                |> String.endsWith locale.negativeSuffix
```
I don't think we need either the `|>` syntax for a single function call, and we don't need to make 3 expressions to accomplish what we need:

```suggestion
            String.startsWith locale.negativePrefix value && String.endsWith locale.negativeSuffixMatch value
```

praveentiru replied:
Agreed and I have implemented the same.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,61 @@ parse locale original =
+                |> String.split locale.decimalSeparator
+                |> List.map (Regex.replace chrs (\_ -> ""))
+
+        sumNumber : Float -> Float -> Float -> Float
+        sumNumber i d pow =
+            if isNegative then
+                -1 * (i + d / (10 ^ pow))
+
+            else
+                i + d / (10 ^ pow)
```
This function has two responsibilities, coupling the logic too much. Let's think about a `toFloat` and we multiply by `-1` at the end of the `parseSting` chain. Also, why do we multiply by `-1` here and then, again, in line 432? This logic is confusing.

Can we rename `i` and `d`? They are not meaningful variable names. What about `integers` and `decimals`? And talking about meaningful names, can you explain in plain English what `sumNumber` does? I would like to come up with a better name for this function.

Finally, I would treat `pow` as part of the logic of this function, and not as something calculated elsewhere.

Keeping all that in mind, I think of a refactor `sumNumber` as `toFloat : (String, String) -> Float`, something among these lines:

```elm
toFloat parts =
    let
        pow : Float
        pow =
             parts
                |> Tuple.second
                |> String.length

        decimals : Float
        decimals =
             parts
                |> Tuple.second
                |> String.toFloat
                |> Maybe.withDefault 0.0

        integers : Float
        integers =
             parts
                |> Tuple.first
                |> String.toFloat
                |> Maybe.withDefault 0.0 
    in
    integers + decimals / (10 ^ pow) 
```

praveentiru replied:
Added function with name `joinParts`. It is opposite of `splitInParts`. It also performs an opposite function. It takes integer and decimal strings and returns `Maybe Float`. Rules are encoded in tests in comments.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +380,61 @@ parse locale original =
+            else
+                i + d / (10 ^ pow)
+    in
+    case splitValue value of
```
As we expect to return a `Float`, I would avoid working with a list — a data type that could have none, one, two or more items. I would use a tuple (guaranteed two items: integers, and decimals in our case) instead.

Or even a custom type `{ decimals: Int, integers: Int}`. Let's [make impossible states impossible](https://www.youtube.com/watch?v=IcgmSRJHu_8) — but a tuple `(Float, Float)` works fine here : )

praveentiru replied:
Agreed. `joinParts` uses `(String, String)` as input to ensure that only two parts are passed as input.

praveentiru committed:
Change summary:

1. Added new function joinParts to build number from parts
2. Cleaned up code in parseString as per feedback

praveentiru committed:
Make changes to parseString function with better naming as advised

praveentiru committed:
Using elm's negate function rather than custom logic

praveentiru committed:
Merge branch 'parse-string' of github.com:praveentiru/elm-format-number into parse-string

praveentiru commented:
@cuducos I have never used rebase on command line. I believe I have messed it up. Please squash the commits and merge if all changes are ok.

cuducos commented on src/FormatNumber.elm:
```diff
@@ -139,5 +142,33 @@ import FormatNumber.Stringfy exposing (stringfy)
+    parse { base | thousandSeparator = "," } "100,000.345"
+    --> Just 100000.345
+
+    parse { base | thousandSeparator = ",", negativePrefix = "-" } "-100#000"
```
Why the `#` in the `-100#000` example? If `#` is not the thousand separators, this example seems wrong, right?

cuducos commented on src/FormatNumber.elm:
```diff
@@ -1,4 +1,7 @@
-module FormatNumber exposing (format)
+module FormatNumber exposing
+    ( format
+    , parse
+    )
```
Does Elm Format do this? I thought it would be better like:

```suggestion
module FormatNumber exposing (format, parse)
```

praveentiru replied:
Yes it is from elm format. I am new to Elm so, just used it to format before commit.

cuducos replied:
Excellent then : )

cuducos commented on src/FormatNumber.elm:
```diff
@@ -139,5 +142,33 @@ import FormatNumber.Stringfy exposing (stringfy)
+    parse { base | thousandSeparator = "," } "100,000.345"
+    --> Just 100000.345
+
+    parse { base | thousandSeparator = ",", negativePrefix = "-" } "-100#000"
+    --> Just -100000
```
Why the `#` in the `-100#000` example? If `#` is not a thousand separators, this example seems wrong, right?

praveentiru replied:
Yes. It is wrong. But, this is the problem with current approach of cleaning integer part of non-digits. Thousand separator does not have any value in parsing.

cuducos replied:
We can opt for a strategy to validate every every non-digit is one of the expected tokens (negative prefix or suffix, zero prefix or suffix, separators):

1. Split the value by the regex `\d+`
2. Check if the resulting list contains only expected tokens

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -211,6 +214,67 @@ getDecimals locale digits =
+   joinParts ("10000", "24,3")
+   --> Nothing
+
+-}
+
+
+joinParts : ( String, String ) -> Maybe Float
```
```suggestion
-}
joinParts : ( String, String ) -> Maybe Float
```

praveentiru replied:
Additional new lines were from Elm format. I ran it multiple time to make sure. I don't know if it is some weird bug there.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -211,6 +214,67 @@ getDecimals locale digits =
 
 
+
+{- Joins parts into `Maybe Float`. It expects parts to be strings with only digits.
```
I may be wrong, but I think the style suggests wrapping at 80 chars.

```suggestion
{- Joins parts into `Maybe Float`. It expects parts to be strings with only
digits.
```

praveentiru replied:
ok. I will make this change

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -211,6 +214,67 @@ getDecimals locale digits =
+                |> String.toFloat
+                |> Maybe.map (\n -> n / (10 ^ decimalCount))
+    in
+    case parts of
+        ( "", "" ) ->
+            Nothing
+
+        ( _, "" ) ->
+            integers
+
+        ( "", _ ) ->
+            decimalValue
+
+        _ ->
+            Maybe.map2 (+) integers decimalValue
```
Nice :heart_eyes_cat:

praveentiru replied:
Thanks!

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +442,54 @@ parse locale original =
             }
+
+
+{-| Given a `Locale` parses a `String` into a `Maybe Float`
+-}
```
We need some examples here, both for documentation and for tests ; )

praveentiru replied:
Reason I did not add any tests here is because parse function in FormatNumber.elm is just calls this function and does nothing else. I thought it would be repetition.

cuducos replied:
> parse function in FormatNumber.elm is just calls this function

So, what about moving this function to `FormatNumber.elm`?

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +442,54 @@ parse locale original =
+    let
+        isNegative : Bool
+        isNegative =
+            String.startsWith locale.negativePrefix value
+                && String.endsWith
+                    locale.negativeSuffix
+                    value
```
Is this formatting what Elm Format suggests? Weird, but if it is Elm Format, we accept it hahaha…

praveentiru replied:
It is from Elm format.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +442,54 @@ parse locale original =
+        onlyDigits = 
+            Regex.replace notDigits (\_ -> "")
+
+        splitValue : String -> List String
```
I still think we should use `String -> (String, String)` here.

praveentiru replied:
If I change this, then I can move entire logic of `joinParts` into the `parseString`. I like the current structure as it cleanly segregates that out of `parseString` and allow me to test it separately.

Let me know if you still want it changed. I will do it.

cuducos replied:
I think making impossible states impossible worth the job : )

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +442,54 @@ parse locale original =
+            if isNegative then
+                ( integers, decimals )
+                    |> joinParts
+                    |> Maybe.map (negate)
```
Do we need this parenthesis? 

```suggestion
                    |> Maybe.map negate
```

praveentiru replied:
Agreed. Removing parenthesis.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -378,3 +442,54 @@ parse locale original =
+            if isNegative then
+                ( integers, "" )
+                    |> joinParts
+                    |> Maybe.map (negate)
```
Do we need this parethesis?

```suggestion
                    |> Maybe.map negate
```

praveentiru replied:
Agreed. Removing parenthesis.

praveentiru commented:
@cuducos I believe there is a major flaw that I was not thinking about that is exposed by example: `parse { base | thousandSeparator = ",", negativePrefix = "-" } "-100#000"`

I have not built any code to validate the string against locale before parsing. I am working on a function to do the same. Will take little time but, I am hoping that would be a better approach. I will create a separate branch for the same so, we are able to evaluate that approach while building on the current code.

praveentiru committed:
Minor edits based on reviewer feedback
